### PR TITLE
Auth cookie domain

### PR DESCRIPTION
**Pull request by pmahindrakar-oss**
# TL;DR
Allow user of flyteadmin to specify the cookie settings of there flyteadmin user auth cookies
Allowing subdomain access by setting the domain field in the cookie
and allowing to set SameSite preference to allow firstparty or thirdparty cookies to be sent from the browser.

Mode changes are applicable for only the following cookies 
* flyte_at
* flyte_idt
* flyte_user_info

One thing i noticed is that dot prefix gets subsituted by the browser seems like .

```
A cover domain should not contain a leading dot, like in .cats.com; if it does, the client should remove the leading dot.
```
source : http://bayou.io/draft/cookie.domain.html

So  i didn't find any behavior change with domain setting policy for localhost.

#### With LAX mode for auth cookies

Screenshot 2022-06-07 at 5 10 37 PM

#### With Strict mode for auth cookies

Screenshot 2022-06-07 at 5 09 17 PM

## Type
 - [ ] Bug Fix
 - [X] Feature
 - [ ] Plugin

## Are all requirements met?

 - [X] Code completed
 - [X] Smoke tested
 - [X] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
_Remove the '*fixes*' keyword if there will be multiple PRs to fix the linked issue_

fixes https://github.com/flyteorg/flyte/issues/2596

## Follow-up issue
_NA_

### CODE REVIEW

File: auth/cookie_manager.go [hunk 1]
```diff
@@ -194,3 +199,10 @@ func (c CookieManager) DeleteCookies(ctx context.Context, writer http.ResponseWr
 	http.SetCookie(writer, getLogoutAccessCookie())
 	http.SetCookie(writer, getLogoutRefreshCookie())
 }
+
+func (c CookieManager) getCookieDomain(request *http.Request) string {
+	if !c.coverSubDomains {
+		return ""
+	}
+	return fmt.Sprintf(".%s", request.URL.Hostname())
```
Comment (EngHabu) [inline auth/cookie_manager.go:207]: nice! I guess this works because this is an HTTP request not a GRPC request...

Have you tried it behind an ingress? mind doing that?

File: auth/config/config.go [hunk 2]
```diff
@@ -212,8 +217,14 @@ type UserAuthConfig struct {
...
 	// Secret names, defaults are set in DefaultConfig variable above but are possible to override through configs.
-	CookieHashKeySecretName  string `json:"cookieHashKeySecretName" pflag:",OPTIONAL: Secret name to use for cookie hash key."`
-	CookieBlockKeySecretName string `json:"cookieBlockKeySecretName" pflag:",OPTIONAL: Secret name to use for cookie block key."`
+	CookieHashKeySecretName  string          `json:"cookieHashKeySecretName" pflag:",OPTIONAL: Secret name to use for cookie hash key."`
+	CookieBlockKeySecretName string          `json:"cookieBlockKeySecretName" pflag:",OPTIONAL: Secret name to use for cookie block key."`
+	CookieSetting            *CookieSettings `json:"cookieSetting" pflag:", settings used by cookies created for user auth"`
+}
+
+type CookieSettings struct {
+	SameSite        http.SameSite `json:"sameSite" pflag:",OPTIONAL: Allows you to declare if your cookie should be restricted to a first-party or same-site context."`
```
Comment (EngHabu) [inline auth/config/config.go:226]: Does this parse well from a yaml config?

**Review by EngHabu — COMMENTED** [review]
Looks great! thank you for doing that! a couple of nits

**Review by ursucarina — APPROVED** [review]

File: auth/config/config.go [hunk 3]
```diff
@@ -212,8 +217,14 @@ type UserAuthConfig struct {
...
+	CoverSubdomains bool          `json:"coverSubDomains" pflag:",OPTIONAL: Allow subdomain access to the created cookies by setting the domain attribute."`
```
[the remark this answers is not in the document]
Comment (pmahindrakar-oss) [reply auth/config/config.go]: Done. Thanks

File: auth/cookie_manager.go (hunk 1 above)
Comment (pmahindrakar-oss) [reply auth/cookie_manager.go:222]: Tested this both with grpc endpoints through flytectl and through flyteconsole and no issues were discovered

File: auth/config/config.go (hunk 2 above)
Comment (pmahindrakar-oss) [reply auth/config/config.go]: It does if we pass int values. But added an enum wrapper for good UX.

**Review by EngHabu — APPROVED** [review]
